### PR TITLE
Adding new mode to deal with non-prompt lepton ntuple

### PR DESCRIPTION
**Pull request by jeongsumin**
This pr is the first version that included non-prompt lepton and should be discussed about NPL logic.

### CODE REVIEW

File: TopAnalysis/interface/FCNCTriLeptonCppWorker.h [hunk 1]
```diff
@@ -22,7 +22,7 @@ class FCNCTriLeptonCppWorker {
   ~FCNCTriLeptonCppWorker() = default;
 
   void setMuons(TRAF pt, TRAF eta, TRAF phi, TRAF mass, TRAI charge,
-                TRAF relIso, TRAB isTight, TRAB isGlobal, TRAB isPFcand, TRAB isTracker);
+                TRAF relIso, TRAB isLoose, TRAB isTight, TRAB isGlobal, TRAB isPFcand, TRAB isTracker);
```
Comment (jhgoh) [inline TopAnalysis/interface/FCNCTriLeptonCppWorker.h]: is there Muons_isLoose in the NanoAOD? you have to use isGlobal, isPFcan and isTracker to define isLoose

File: TopAnalysis/interface/FCNCTriLeptonCppWorker.h [hunk 2]
```diff
@@ -22,7 +22,7 @@ class FCNCTriLeptonCppWorker {
...
   void setElectrons(TRAF pt, TRAF eta, TRAF phi, TRAF mass, TRAI charge,
                     TRAF relIso, TRAI id, TRAF dEtaSC, TRAF eCorr);
```
Comment (jhgoh) [inline TopAnalysis/interface/FCNCTriLeptonCppWorker.h]: need lots of modification in the setElectron to invert isolation. electrion id code is just for the isolated electrons.

File: TopAnalysis/interface/FCNCTriLeptonCppWorker.h [hunk 3]
```diff
@@ -32,6 +32,16 @@ class FCNCTriLeptonCppWorker {
...
+  float get_LeadingMuon_eta() const { return out_LeadingMuon_p4[1]; }
+  float get_LeadingMuon_phi() const { return out_LeadingMuon_p4[2]; }
+  float get_LeadingMuon_mass() const { return out_LeadingMuon_p4[3]; }
+
+  float get_LeadingElectron_pt() const { return out_LeadingElectron_p4[0]; }
+  float get_LeadingElectron_eta() const { return out_LeadingElectron_p4[1]; }
+  float get_LeadingElectron_phi() const { return out_LeadingElectron_p4[2]; }
+  float get_LeadingElectron_mass() const { return out_LeadingElectron_p4[3]; }
+
   float get_LeadingLepton_pt() const { return std::max(std::max(get_Lepton1_pt(), get_Lepton2_pt()), get_Lepton3_pt()); }
```
Comment (jhgoh) [inline TopAnalysis/interface/FCNCTriLeptonCppWorker.h]: change this to std::max(out_LeadingMuon_p4[0], out_LeadingElectron_p4[0]);

File: TopAnalysis/python/postprocessing/fcncTriLepton.py [hunk 4]
```diff
@@ -37,10 +37,11 @@ def beginFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
         self.out.branch("Lepton1_pdgId", "I")
         self.out.branch("Lepton2_pdgId", "I")
         self.out.branch("Lepton3_pdgId", "I")
-        self.out.branch("LeptonTotal_mass", "F")
-        self.out.branch("LeptonTotal_pt", "F")
-        self.out.branch("LeptonWandZ_deltaPhi", "F")
-        self.out.branch("LeptonWandZ_deltaR", "F")
+	self.out.branch("LeadingLepton_pt", "F")
```
Comment (jhgoh) [inline TopAnalysis/python/postprocessing/fcncTriLepton.py]: Keep the same indentation NEVER USE TAB character.

File: TopAnalysis/python/postprocessing/fcncTriLepton.py [hunk 5]
```diff
@@ -69,7 +70,7 @@ def initReaders(self,tree):
         objName = "Muon"
         setattr(self, "b_n%s" % objName, tree.valueReader("n%s" % objName))
         for varName in ["pt", "eta", "phi", "mass", "charge",
-                        "pfRelIso04_all", "tightId", "isGlobal", "isPFcand", "isTracker"]:
+                        "pfRelIso04_all", "looseId", "tightId", "isGlobal", "isPFcand", "isTracker"]:
```
Comment (jhgoh) [inline TopAnalysis/python/postprocessing/fcncTriLepton.py]: no "looseId" for the muons

File: TopAnalysis/python/postprocessing/fcncTriLepton.py [hunk 6]
```diff
@@ -83,7 +84,7 @@ def initReaders(self,tree):
                                  self.b_Electron_pfRelIso03_all, getattr(self, 'b_Electron_%s' % self.eleIdName),
                                  self.b_Electron_deltaEtaSC, self.b_Electron_eCorr)
         self.worker.setMuons(self.b_Muon_pt, self.b_Muon_eta, self.b_Muon_phi, self.b_Muon_mass, self.b_Muon_charge,
-                             self.b_Muon_pfRelIso04_all,self.b_Muon_tightId, self.b_Muon_isGlobal, self.b_Muon_isPFcand, self.b_Muon_isTracker)
+                             self.b_Muon_pfRelIso04_all, self.b_Muon_looseId, self.b_Muon_tightId, self.b_Muon_isGlobal, self.b_Muon_isPFcand, self.b_Muon_isTracker)
```
Comment (jhgoh) [inline TopAnalysis/python/postprocessing/fcncTriLepton.py]: no loose id for the muons

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 7]
```diff
@@ -36,13 +40,14 @@ void FCNCTriLeptonCppWorker::setElectrons(TRAF pt, TRAF eta, TRAF phi, TRAF mass
 }
 
 void FCNCTriLeptonCppWorker::setMuons(TRAF pt, TRAF eta, TRAF phi, TRAF mass, TRAI charge,
-                                      TRAF relIso, TRAB isTight, TRAB isGlobal, TRAB isPFcand, TRAB isTracker) {
+                                      TRAF relIso, TRAB isLoose, TRAB isTight, TRAB isGlobal, TRAB isPFcand, TRAB isTracker) {
```
Comment (jhgoh) [inline TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: no isLoose.

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 8]
```diff
@@ -36,13 +40,14 @@ void FCNCTriLeptonCppWorker::setElectrons(TRAF pt, TRAF eta, TRAF phi, TRAF mass
...
   in_Muons_p4[0] = pt;
   in_Muons_p4[1] = eta;
   in_Muons_p4[2] = phi;
   in_Muons_p4[3] = mass;
   in_Muons_charge = charge;
   in_Muons_relIso = relIso;
+  in_Muons_isLoose = isLoose;
```
Comment (jhgoh) [inline TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: no isLoose.

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 9]
```diff
@@ -126,6 +143,21 @@ bool FCNCTriLeptonCppWorker::isVetoElectron(const unsigned i) const {
...
 }
+//nonprompt electrons
+bool FCNCTriLeptonCppWorker::isNPElectron(const unsigned i) const {
+  const double pt = in_Electrons_p4[0]->At(i);
+  const double eta = in_Electrons_p4[1]->At(i);
+  const double scEta = eta + in_Electrons_dEtaSC->At(i);
+  if ( pt < minElectronPt_ or std::abs(eta) > maxElectronEta_ ) return false;
+  if ( std::abs(scEta) > 1.4442 and std::abs(scEta) < 1.566 ) return false;
+  //nanoAOD object -> Electron_cutBased_Sum16 0:fail, 1:veto, 2:loose, 3:medium, 4:tight
+  if ( in_Electrons_id->At(i) != 2 ) return false;
```
Comment (jhgoh) [inline TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: 1. you are asking electron id == 2. this means you are rejecting medium, tight electrons.
2. need special electron id without isolation... no?

File: TopAnalysis/interface/FCNCTriLeptonCppWorker.h (hunk 3 above)
Comment (jeongsumin) [reply TopAnalysis/interface/FCNCTriLeptonCppWorker.h]: Should we see the leading muon/electron regardless of channel?
In NTU presentation, they drew the leading muon when the 3lepton has >= 2mu and leading electron when the 3 lepton has >=2el.
Or do you mean return should be changed like "std::max(out_Lepton2_p4[0], out_Lepton3_p4[0])" at the line 41 to 43..?

File: TopAnalysis/python/postprocessing/fcncTriLepton.py [hunk 10]
```diff
@@ -63,7 +64,7 @@ def initReaders(self,tree):
         objName = "Electron"
         setattr(self, "b_n%s" % objName, tree.valueReader("n%s" % objName))
         for varName in ["pt", "eta", "phi", "mass", "charge",
-                        "pfRelIso03_all", self.eleIdName, "deltaEtaSC", "eCorr",]:
+                        "pfRelIso03_all", self.eleIdName, "deltaEtaSC", "eCorr", "vidNestedWPBitmapSum16",]:
```
Comment (jhgoh) [inline TopAnalysis/python/postprocessing/fcncTriLepton.py]: Should be able to change this vid bitmap name - to run on 2017+ datasets for the future.

File: TopAnalysis/python/postprocessing/fcncTriLepton.py [hunk 11]
```diff
@@ -116,8 +117,16 @@ def analyze(self, event):
 fcnc_ElElEl_2016 = lambda : FCNCTriLepton(mode="ElElEl", eleId="cutBased_Sum16")
 fcnc_ElMuMu_2016 = lambda : FCNCTriLepton(mode="ElMuMu", eleId="cutBased_Sum16")
 fcnc_MuElEl_2016 = lambda : FCNCTriLepton(mode="MuElEl", eleId="cutBased_Sum16")
+fcnc_NPLMuMuMu_2016 = lambda : FCNCTriLepton(mode="NPLMuMuMu", eleId="cutBased_Sum16")
```
Comment (jhgoh) [inline TopAnalysis/python/postprocessing/fcncTriLepton.py]: (related with the previous comment - say we have one more argument such as eleVidMap="vid.....Sum16"

Comment (jhgoh) [reply TopAnalysis/python/postprocessing/fcncTriLepton.py]: BTW, better to comment out at this moment as we decided not to produce NPL's at this moment.

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 12]
```diff
@@ -23,7 +27,7 @@ typedef FCNCTriLeptonCppWorker::TRAI TRAI;
 typedef FCNCTriLeptonCppWorker::TRAB TRAB;
 
 void FCNCTriLeptonCppWorker::setElectrons(TRAF pt, TRAF eta, TRAF phi, TRAF mass, TRAI charge,
-                                              TRAF relIso, TRAI id, TRAF dEtaSC, TRAF eCorr) {
+                                          TRAF relIso, TRAI id, TRAF dEtaSC, TRAF eCorr, TRAI vidNestedWPBitmapSum16) {
```
Comment (jhgoh) [inline TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: Better to rename vidNestedWPBitmapSum16 -> vidBitmap or similar.

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 13]
```diff
@@ -126,6 +143,22 @@ bool FCNCTriLeptonCppWorker::isVetoElectron(const unsigned i) const {
...
+  const double eta = in_Electrons_p4[1]->At(i);
+  const double scEta = eta + in_Electrons_dEtaSC->At(i);
+  if ( pt < minElectronPt_ or std::abs(eta) > maxElectronEta_ ) return false;
+  if ( std::abs(scEta) > 1.4442 and std::abs(scEta) < 1.566 ) return false;
+  //nanoAOD object -> Electron_vidNestedWPBitmapSum16 (total 10 cuts in here by bitmap, each cut has 3 bit)
+  const char vidNestedWPBitmapSum16 = in_Electrons_vidNestedWPBitmapSum16->At(i);
+  //if ( ! ( vidNestedWPBitmapSum16 > ... ) ) return false;
+  //Isolation criteria from signal electrons are inverted in NPL
+  if ( std::abs(eta) <= 1.479 ) { if ( in_Electrons_relIso->At(i) < 0.0588 ) return false; }
+  if ( std::abs(eta) > 1.479 ) { if ( in_Electrons_relIso->At(i) < 0.0571 ) return false; }
```
Comment (jhgoh) [inline TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: isolation cuts are already in the vid.

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 14]
```diff
@@ -171,24 +204,34 @@ bool FCNCTriLeptonCppWorker::analyze() {
   // Select leptons
   std::vector<int> muonIdxs;
   std::vector<int> electronIdxs;
+  std::vector<int> NPmuonIdxs;
```
Comment (jhgoh) [inline TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: naming convention. never start with capitals for ordinary variables.

File: TopAnalysis/src/FCNCTriLeptonCppWorker.cc [hunk 15]
```diff
@@ -222,6 +265,24 @@ bool FCNCTriLeptonCppWorker::analyze() {
       out_Lepton3_pdgId = -11*in_Electrons_charge->At(electronIdxs[1]);
     }
   }
+  else if ( actualMode == MODE::NPLMuElEl ) {
```
[the remark this answers is not in the document]
Comment (jhgoh) [reply TopAnalysis/src/FCNCTriLeptonCppWorker.cc]: of course needs some additional routines .

**Comment by jeongsumin** [comment]
Since the above modification, the test job is run well.

**Pull request merged**